From 273f8381677921e7452a87cd8718cd9e21d273d2 Mon Sep 17 00:00:00 2001
From: Wen He <wen.he_1@nxp.com>
Date: Thu, 13 Dec 2018 16:28:14 +0800
Subject: [PATCH 401/706] drm: hdp: auto sync pixel clock from different
 resolution

This patch implement a pixel clock driver for DP.

Signed-off-by: Wen He <wen.he_1@nxp.com>
(cherry picked from commit 47e934315fcd34b492e69fdbe922b964ae666bab)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 .../arm64/boot/dts/freescale/fsl-ls1028a.dtsi |  3 +-
 drivers/gpu/drm/imx/hdp/imx-hdp.c             | 55 +++++++++++++++----
 drivers/gpu/drm/imx/hdp/imx-hdp.h             |  8 +++
 3 files changed, 55 insertions(+), 11 deletions(-)

diff --git a/arch/arm64/boot/dts/freescale/fsl-ls1028a.dtsi b/arch/arm64/boot/dts/freescale/fsl-ls1028a.dtsi
index 754d30f5baee..3eb4e2b048ae 100755
--- a/arch/arm64/boot/dts/freescale/fsl-ls1028a.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-ls1028a.dtsi
@@ -824,7 +824,8 @@
 
 	edp: phy@f200000 {
 		compatible = "fsl,ls1028a-dp";
-		reg = <0x0 0xf200000 0x0 0x100000>;
+		reg = <0x0 0xf1f0000 0x0 0xffff>, /* Multimedia PLL Register */
+		    <0x0 0xf200000 0x0 0xfffff>; /* HDP Register*/
 		interrupts = <0 221 0x4>;
 		clocks = <&coreclk>,
 			 <&coreclk>,
diff --git a/drivers/gpu/drm/imx/hdp/imx-hdp.c b/drivers/gpu/drm/imx/hdp/imx-hdp.c
index 8819fb9d911f..b3fc6413ba60 100644
--- a/drivers/gpu/drm/imx/hdp/imx-hdp.c
+++ b/drivers/gpu/drm/imx/hdp/imx-hdp.c
@@ -233,6 +233,40 @@ void imx8mq_phy_reset(sc_ipc_t ipcHndl, struct hdp_mem *mem, u8 reset)
 
 	return;
 }
+
+#else
+
+static void ls1028a_pixel_link_mux(state_struct *state,
+				   struct drm_display_mode *mode)
+{
+	struct imx_hdp *hdp = state_to_imx_hdp(state);
+	u32 val;
+
+	if (mode->hdisplay == 3840
+	    && mode->vdisplay == 2160)
+		val = 0x0402002c;
+	else if (mode->hdisplay == 1920
+		 && mode->vdisplay == 1080)
+		val = 0x1002002c;
+	else if (mode->hdisplay == 1280
+		 && mode->vdisplay == 720)
+		val = 0x2002002c;
+	else if (mode->hdisplay == 720
+		 && mode->vdisplay == 480)
+		val = 0x5802002c;
+	else
+		/* Set deafault pixel clock to 1080p*/
+		val = 0x1002002c;
+
+	writel(val, hdp->mem.ss_base + CSR_PLLDIG_PLLDV);
+
+	/* Recommend register set for PLL Divider */
+	writel(0x40000000, hdp->mem.ss_base + CSR_PLLDIG_PLLFM);
+	writel(0x40030000, hdp->mem.ss_base + CSR_PLLDIG_PLLFD);
+	writel(0x44000000, hdp->mem.ss_base + CSR_PLLDIG_PLLCAL1);
+	writel(0x0005002b, hdp->mem.ss_base + CSR_PLLDIG_PLLCAL2);
+}
+
 #endif
 
 static const struct of_device_id scfg_device_ids[] = {
@@ -971,6 +1005,7 @@ static struct hdp_ops ls1028a_dp_ops = {
 	.get_edid_block = dp_get_edid_block,
 	.get_hpd_state = dp_get_hpd_state,
 	.phy_reset = ls1028a_phy_reset,
+	.pixel_link_mux = ls1028a_pixel_link_mux,
 	.clock_init = imx8qm_clock_init,
 	.ipg_clock_enable = imx8qm_ipg_clock_enable,
 	.ipg_clock_disable = imx8qm_ipg_clock_disable,
@@ -1083,23 +1118,23 @@ static int imx_hdp_imx_bind(struct device *dev, struct device *master,
 	}
 
 	mutex_init(&hdp->mem.mutex);
-	/* register map */
+
+	/* PLL register map */
 	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	hdp->mem.ss_base = devm_ioremap_resource(dev, res);
+	if (IS_ERR(hdp->mem.ss_base)) {
+		dev_err(dev, "Failed to get Multimedia PLL base register\n");
+		return -EINVAL;
+	}
+
+	/* register map */
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 1);
 	hdp->mem.regs_base = devm_ioremap_resource(dev, res);
 	if (IS_ERR(hdp->mem.regs_base)) {
 		dev_err(dev, "Failed to get HDP CTRL base register\n");
 		return -EINVAL;
 	}
 
-	if (!of_device_is_compatible(dev->of_node, "fsl,ls1028a-dp")) {
-		res = platform_get_resource(pdev, IORESOURCE_MEM, 1);
-		hdp->mem.ss_base = devm_ioremap_resource(dev, res);
-		if (IS_ERR(hdp->mem.ss_base)) {
-			dev_err(dev, "Failed to get HDP CRS base register\n");
-			return -EINVAL;
-		}
-	}
-
 	hdp->is_edp = of_property_read_bool(pdev->dev.of_node, "fsl,edp");
 
 	hdp->no_edid = of_property_read_bool(pdev->dev.of_node, "fsl,no_edid");
diff --git a/drivers/gpu/drm/imx/hdp/imx-hdp.h b/drivers/gpu/drm/imx/hdp/imx-hdp.h
index 62aa7e448db1..7aba4da9ba9e 100644
--- a/drivers/gpu/drm/imx/hdp/imx-hdp.h
+++ b/drivers/gpu/drm/imx/hdp/imx-hdp.h
@@ -51,6 +51,14 @@
 #define VIC_MODE_96_50Hz 96
 #define VIC_MODE_97_60Hz 97
 
+#ifdef CONFIG_ARCH_LAYERSCAPE
+#define CSR_PLLDIG_PLLDV 0x28
+#define CSR_PLLDIG_PLLFM 0x2c
+#define CSR_PLLDIG_PLLFD 0x30
+#define CSR_PLLDIG_PLLCAL1 0x38
+#define CSR_PLLDIG_PLLCAL2 0x3c
+#endif
+
 /**
  * imx_hdp_call - Calls a struct imx hdp_operations operation on
  *	an entity
-- 
2.17.1

